lib/active_merchant/billing/gateways/ct_payment.rb

```diff
@@ -0,0 +1,297 @@
+        post[:OriginalAuthorizationNumber] = authorization_number
+        post[:OriginalInvoiceNumber] = invoice_number
+
+        response = commit('completion', post)
+        send_ack(response.authorization)
+        response
+      end
+
+      def refund(money, authorization, options={})
+        requires!(options, :order_id)
+        post = {}
+        add_invoice(post, money, options)
+        add_money(post, money)
+        add_customer_data(post, options)
+        transaction_number, _, invoice_number = split_authorization(authorization)
+        post[:OriginalTransactionNumber] = transaction_number
+        post[:OriginalInvoiceNumber] = invoice_number
+
+        response = commit('refundWithoutCard', post)
+        send_ack(response.authorization)
+        response
+      end
+
+      def credit(money, payment, options={})
+        post = {}
+        add_terminal_number(post, options)
+        add_money(post, money)
+        add_operator_id(post, options)
+        add_invoice(post, money, options)
+        add_payment(post, payment)
+        add_address(post, payment, options)
+        add_customer_data(post, options)
+
+        response = payment.is_a?(String) ? commit('refundWithToken', post) : commit('refund', post)
+        send_ack(response.authorization)
+        response
+      end
+
+      def void(authorization, options={})
+        post = {}
+        post[:InputType] = 'I'
+        post[:LanguageCode] = 'E'
+        transaction_number, _, invoice_number =
+          split_authorization(authorization)
+        post[:OriginalTransactionNumber] = transaction_number
+        post[:OriginalInvoiceNumber] = invoice_number
+        add_operator_id(post, options)
+        add_customer_data(post, options)
+
+        response = commit('void', post)
+        send_ack(response.authorization)
+        response
+      end
+
+      def verify(credit_card, options={})
+        post = {}
+        add_terminal_number(post, options)
+        add_operator_id(post, options)
+        add_invoice(post,0, options)
+        add_payment(post, credit_card)
+        add_customer_data(post, options)
+
+        response = commit('verifyAccount', post)
+        send_ack(response.authorization)
+        response
+      end
+
+      def store(credit_card, options={})
+        requires!(options, :email)
+        post = {}
+        post[:LanguageCode] = 'E'
+        post[:Name] = credit_card.name.rjust(50, ' ')
+        post[:Email] = options[:email].rjust(240, ' ')
+        add_operator_id(post, options)
+        add_payment(post, credit_card)
+        add_customer_data(post, options)
+
+        response = commit('recur/AddUser', post)
+        send_ack(response.authorization)
+        response
+      end
+
+      def supports_scrubbing?
+        true
+      end
+
+      def scrub(transcript)
+        transcript.
+          gsub(%r((&?auth-api-key=)[^&]*)i, '\1[FILTERED]').
+          gsub(%r((&?token:)[^&]*)i, '\1[FILTERED]').
```

Comment:
I'm confused about how this scrub method works. `token` and `cardNumber` both look as if they'd be in the base64-encoded part of the post, so I'm not clear how this filter would work in general. But beyond that, I'm also unclear why they end with `:` instead of `=`, since they're done as POST params. Could you walk me through what I'm missing here?

Reply:
Okay, I think I get what's going on after reading the tests: we're scrubbing the value they echo back in the response, not our submission. Ditto for the token.

_Should_ we scrub the base64 somehow? Do we do this for any other gateway? I'm asking because I'd generally consider preserving base64 strings a PAN leak, but I also don't want to remove all the debugging info…

Reply:
Good catch, scrubbing the payload.